core/src/test/java/org/elasticsearch/index/mapper/DynamicMappingIntegrationIT.java

```diff
@@ -32,6 +34,18 @@
 
 public class DynamicMappingIntegrationIT extends ESIntegTestCase {
 
+    public void testDynamicDisabled() throws IOException {
+        Settings settings = Settings.settingsBuilder().put("index.mapper.dynamic", false).build();
```

Comment:
This setting is per index and it is honored every time (no bug). What's not working is when you set this setting for all indices.
To do this you can override nodeSettings in your test: 

    protected Settings nodeSettings(int nodeOrdinal) {
        return settingsBuilder().put(super.nodeSettings(nodeOrdinal))
            .put("index.mapper.dynamic", "false")
            .build();
    }